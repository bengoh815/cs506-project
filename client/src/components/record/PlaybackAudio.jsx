/******************************************************************************
 * Filename: PlaybackAudio.jsx
 * Purpose:  Renders a button that plays or pauses an audio recording.
 * Author:   Victor Nguyen
 *
 * Description:
 * This component renders a button that plays or pauses an audio recording.
 *
 * Usage:
 * To use this component, import it into the desired file and render it.
 *
 * Note: Additional functionality to be added in the future (ex. seek bar).
 *
 ******************************************************************************/

import React, { useState, useRef } from "react";
import { Button } from "react-bootstrap";

const Playback = ({ audioSrc }) => {
  // Create a reference to the audio element
  const audioRef = useRef();

  // State to store the label of the play/pause button
  const [playPauseLabel, setPlayPauseLabel] = useState("Play Recording");

  /**
   * Plays or pauses the audio.
   */
  const handlePlayPause = () => {
    if (audioRef.current.paused) {
      setPlayPauseLabel("Pause");
      audioRef.current.play();

      // When the recording ends, change the label of the play/pause button
      audioRef.current.onended = () => setPlayPauseLabel("Play Recording");
    } else {
      setPlayPauseLabel("Resume");
      audioRef.current.pause();
    }
  };

  return (
    <div>
      <audio src={audioSrc} ref={audioRef} controls hidden />
      <Button onClick={handlePlayPause}>{playPauseLabel}</Button>
    </div>
  );
};

export default Playback;
