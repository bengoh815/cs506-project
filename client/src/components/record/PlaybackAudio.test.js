/******************************************************************************
 * Filename: PlaybackAudio.test.js
 * Purpose:  Tests the PlaybackAudio component.
 * Author:   Victor Nguyen
 *
 * Description:
 * This file contains tests for the PlaybackAudio component. The tests ensure that the component renders correctly. A snapshot test is included to catch unintended changes to the component's structure or behavior.
 *
 * Usage:
 * Run the tests using the command `npm test`.
 ******************************************************************************/

import { cleanup, render, screen, fireEvent } from "@testing-library/react";
import PlaybackAudio from "./PlaybackAudio";
import React from "react";

describe("Playback Audio Component", () => {
  beforeEach(() => {
    // Mock the play and pause methods of HTMLMediaElement
    window.HTMLMediaElement.prototype.play = jest.fn();
    window.HTMLMediaElement.prototype.pause = jest.fn();
  });

  // Cleans up the DOM after each test to ensure a clean environment.
  afterEach(() => {
    cleanup();
  });

  // Basic render test
  it("renders the component", () => {
    // Arrange
    render(<PlaybackAudio />);

    // Assert
    expect(screen.getByText("Play Recording")).toBeInTheDocument();
  });

  it("presses the play button", () => {
    // Arrange
    render(<PlaybackAudio />);

    // Act
    fireEvent.click(screen.getByText("Play Recording"));

    // Assert
    expect(screen.getByText("Pause")).toBeInTheDocument();
  });

  it("tests handlePlayPause function when audio is playing", () => {
    // Render the Playback component
    render(<PlaybackAudio />);

    // Act
    fireEvent.click(screen.getByText("Play Recording"));
    fireEvent.click(screen.getByText("Pause"));

    // Assert
    expect(screen.getByText("Resume")).toBeInTheDocument();
  });
});
