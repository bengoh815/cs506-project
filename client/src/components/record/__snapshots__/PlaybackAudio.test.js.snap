// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <audio
    controls={true}
    hidden={true}
  />
  <button
    className="btn btn-primary"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Play Recording
  </button>
</div>
`;
